### PR TITLE
Fix: algos calcs float issues

### PR DESCRIPTION
**Pull request by alexstotsky**
### PR description
- Fixed floating-point issues in algo orders calcs
- Full description https://app.asana.com/0/1125859137800433/1199204485116691

### CODE REVIEW

File: test/lib/twap/events/orders_order_fill.js [hunk 1]
```diff
@@ -55,6 +55,36 @@ describe('twap:events:orders_order_fill', () => {
...
+        remainingAmount: 0.5
+      },
+
+      h: {
+        ...instance.h,
+
+        updateState: (inst, update) => {
+          return new Promise((resolve) => {
+            assert.deepStrictEqual(update, {
+              remainingAmount: 0.3
```
Comment (robertkowalski) [inline test/lib/twap/events/orders_order_fill.js:79]: > 0.5 - 0.2
0.3
> 0.5 - 0.3
0.2

Comment (robertkowalski) [reply test/lib/twap/events/orders_order_fill.js:79]: try 0.3 - 0.2

File: test/lib/twap/util/is_target_met.js [hunk 2]
```diff
@@ -10,4 +10,10 @@ describe('twap:util:is_target_met', () => {
     assert.ok(isTargetMet({ priceTarget: 1, priceDelta: 0.2 }, 1.2))
     assert.ok(!isTargetMet({ priceTarget: 1, priceDelta: 0.2 }, 1.3))
   })
+
+  it('verifies that the price target is met, floats', () => {
+    assert.ok(isTargetMet({ priceTarget: 0.5, priceDelta: 0 }, 0.5))
+    assert.ok(isTargetMet({ priceTarget: 0.5, priceDelta: 0.2 }, 0.7))
```
Comment (robertkowalski) [inline test/lib/twap/util/is_target_met.js:16]: not all floats run into this issue:

```
> 0.5 - 0.2
0.3
> 0.5 - 0.3
0.2
> 0.3 - 0.1
0.19999999999999998
> 0.3 - 0.2
0.09999999999999998
```

File: test/lib/accumulate_distribute/events/orders_order_fill.js [hunk 3]
```diff
@@ -38,11 +38,48 @@ describe('accumulate_distribute:events:orders_order_fill', () => {
...
+
+      h: {
+        ...instance.h,
+
+        updateState: (inst, update) => {
+          return new Promise((resolve) => {
+            assert.deepStrictEqual(update, {
+              currentOrder: 1,
+              ordersBehind: 0,
+              remainingAmount: 0.5
```
Comment (robertkowalski) [inline test/lib/accumulate_distribute/events/orders_order_fill.js:74]: did this run into odd values before the fix (see other comments)

File: lib/ping_pong/util/gen_ping_pong_table.js [hunk 4]
```diff
@@ -23,7 +24,7 @@ const genPingPongTable = (args = {}) => {
   if (orderCount === 1) {
     pingPongTable[preparePrice(pingPrice)] = preparePrice(pongPrice)
   } else {
-    const step = (pingMaxPrice - pingMinPrice) / (orderCount - 1)
+    const step = (nBN(pingMaxPrice).minus(pingMinPrice).toNumber()) / (orderCount - 1)
```
Comment (robertkowalski) [inline lib/ping_pong/util/gen_ping_pong_table.js:27]: > 1.2 / 3
0.39999999999999997

Comment (robertkowalski) [reply lib/ping_pong/util/gen_ping_pong_table.js:27]: not fully fixed here

File: test/lib/accumulate_distribute/events/orders_order_fill.js [hunk 5]
```diff
@@ -38,11 +38,48 @@ describe('accumulate_distribute:events:orders_order_fill', () => {
     o.amount = 40
     assert.strictEqual(o.getLastFillAmount(), 2, 'sanity check failed')
 
-    const i = getInstance({})
+    const i = getInstance({
+      stateParams: {remainingAmount: 0}
```
Comment (robertkowalski) [inline test/lib/accumulate_distribute/events/orders_order_fill.js:42]:
```
  /Users/robert/bitfinex/bfx-hf-algo/test/lib/accumulate_distribute/events/orders_order_fill.js:42:20: A space is required after '{'.
  /Users/robert/bitfinex/bfx-hf-algo/test/lib/accumulate_distribute/events/orders_order_fill.js:42:39: A space is required before '}'.
  /Users/robert/bitfinex/bfx-hf-algo/test/lib/accumulate_distribute/events/orders_order_fill.js:63:26: Unexpected trailing comma.
```

File: test/lib/accumulate_distribute/meta/init_state.js [hunk 6]
```diff
@@ -18,6 +18,17 @@ describe('accumulate_distribute:meta:init_state', () => {
     stubOrderAmountsGen.restore()
   })
 
+  it('initializes order amounts, floats', () => {
+    const stubOrderAmountsGen = sinon.stub(orderAmounts, 'gen').returns([0.001, 0.002])
```
Comment (robertkowalski) [inline test/lib/accumulate_distribute/meta/init_state.js:22]: if `gen.orderAmounts` is stubbed, what float function are we testing here then?

might be missing something, but i think we can delete this test

File: lib/accumulate_distribute/meta/init_state.js [hunk 7]
```diff
@@ -22,9 +23,9 @@ const initState = (args = {}) => {
   const orderAmounts = genOrderAmounts.gen(args)
 
   let totalAmount = 0
-  orderAmounts.forEach(a => { totalAmount += a })
+  orderAmounts.forEach(a => { totalAmount = nBN(totalAmount).plus(a).toNumber() })
 
-  if (Math.abs(totalAmount - amount) > DUST) {
+  if (Math.abs(nBN(totalAmount).minus(amount).toNumber()) > DUST) {
```
Comment (robertkowalski) [inline lib/accumulate_distribute/meta/init_state.js:28]: total amount is just used internally, for a check on dust, no need for that. 

(you can revert this file back)

File: lib/accumulate_distribute/util/gen_order_amounts.js [hunk 8]
```diff
@@ -27,14 +28,14 @@ const generateOrderAmounts = (args = {}) => {
...
-      const remAmount = amount - totalAmount
+      const remAmount = nBN(amount).minus(totalAmount).toNumber()
       const cappedOrderAmount = +prepareAmount(remAmount < 0
         ? Math.max(remAmount, orderAmount)
         : Math.min(remAmount, orderAmount)
       )
 
       orderAmounts.push(cappedOrderAmount)
-      totalAmount += cappedOrderAmount
+      totalAmount = nBN(totalAmount).plus(cappedOrderAmount).toNumber()
```
Comment (robertkowalski) [inline lib/accumulate_distribute/util/gen_order_amounts.js:38]: ok, something is wrong here. if i revert all changes here in the file and run the testsuite, tests still pass for me.

File: lib/accumulate_distribute/util/generate_order.js [hunk 9]
```diff
@@ -4,6 +4,7 @@ const _isFinite = require('lodash/isFinite')
 const _isObject = require('lodash/isObject')
 const { Order } = require('bfx-api-node-models')
 const genCID = require('../../util/gen_client_id')
+const { nBN } = require('@bitfinex/lib-js-util-math')
```
Comment (robertkowalski) [inline lib/accumulate_distribute/util/generate_order.js:7]: this is part the core functionality of the algo. 

would be good to have two tests: one that ensures that we did not break anything and another that verifies the float handling that we added.

**Review by robertkowalski — CHANGES_REQUESTED** [review]
i started with an in-depth review of accumulate-distribute, sadly too many issue found to continue with other algos without making this a very long standing PR where its easy to lose track of important changes and requirements.

suggestion: please open a new PR, just for the corrected accumulate-distribute part. then we get accumulate-distribute ready for shipping and focus on the next algo with another new PR

**Comment by alexstotsky** [comment]
ok, will divide into several PRs, one per strategy

**Pull request closed without merging**